src/core/inventory_component.cpp

```diff
@@ -126,9 +150,45 @@ Opt<NormalItem> InventoryComponent::GetSelectedNormalItem()
     return mSelectedNormalItem;
 }
 
-void InventoryComponent::SetSelectedNormalItem( int32_t Id )
+bool InventoryComponent::SetSelectedNormalItem( int32_t Id, bool force /*= false */ )
 {
-    mSelectedNormalItem = Opt<NormalItem>( dynamic_cast<NormalItem*>( GetItem( Id ).Get() ) );
+    if (force || !mSelectedNormalItem.IsValid() || mSelectedNormalItem->CanSwitch())
+    {
+        if (mSelectedNormalItem.IsValid())
+        {
+            mSelectedNormalItem->Deselected();
+        }
+        mSelectedNormalItem = GetItem( Id );
+        if (mSelectedNormalItem.IsValid())
+        {
+            mSelectedNormalItem->Selected();
+        }
+        return true;
+    }
+    return false;
+}
+
+
+bool InventoryComponent::SwitchToNextItem( ItemType::Type itemType, bool forward /*= true */ )
+{
+    Opt<Item> item( (itemType == ItemType::Normal) ? Opt<Item>(mSelectedNormalItem) : Opt<Item>(mSelectedWeapon) );
+    if (item.IsValid())
+    {
+        auto found = std::find_if( mItems.begin(), mItems.end(), [&]( auto i ) { return i->GetId() == item->GetId(); } );
+        auto i = found;
+        do
+        {
+            if ((forward ? ++i : --i) == mItems.end())
```

Comment:
this is wrong. it will get stuck at begin when switching backwards.
there's no inherent cyclic iterator ( where would you stop, amiright ), i'd suggest using something along the lines of
for( int i = 0, e = mItems.size(); i != e; modify( i ) )
{
   auto const&& item = mItems[ (i + startidx) % e ];
   ... 
}
of course, you'd have to change ItemList_t to a reasonable container from list, but as it already contains Opt<> and not instances of Item, it shouldn't be a problem.